add context manager functionality to ion and ioff

ianhi opened the pull request:
## PR Summary
Closes: https://github.com/matplotlib/matplotlib/issues/17013
Also closes: https://github.com/matplotlib/ipympl/issues/220

This allows the usage of `plt.ioff` and `plt.ion` as context managers:
```python
plt.ion()
with plt.ioff():
    # now interactive mode is off
    fig = plt.figure()
    ax = fig.gca()
```
```python
plt.ioff()
with plt.ion():
    # now interactive mode is on
    fig = plt.figure()
    ax = fig.gca()
```
It also caches the state and returns to the original state on exit:
```python
plt.ion()
with plt.ioff():
    # now interactive mode is off
    fig = plt.figure()
    ax = fig.gca()
assert plt.isinteractive
```
## Documentation - WIP -  advice welcome
I think that this additional behavior ought to be documented but I was a bit unsure how to go about it. In general things in the documentation seem divided into being functions or objects, and this PR puts `ion` and `ioff` somewhere between. So I think it makes sense to keep these on this page: https://matplotlib.org/3.2.1/api/_as_gen/matplotlib.pyplot.html?highlight=pyplot#module-matplotlib.pyplot. But, then they would be mislabelled as Functions, and I'm not sure that they will end up there anyway as that page is autogenerated. I also wasn't sure which docstring to put information about the context manager usage in, I think the most sensible place is in the `__call__` method, but this isn't strictly what is used as the context manager.

The documentation needs be updated at these links:  
https://matplotlib.org/3.2.1/users/shell.html#controlling-interactive-updating  
https://matplotlib.org/3.2.1/tutorials/introductory/usage.html?highlight=usage#what-is-interactive-mode

I'll give this a go but wanted to get input on where to put extended docstrings and how to manage the autogenerating docs. Also, if anyone has any wording suggestions I'd be happy to hear them.


## Tests
There don't seem to be any tests that make use of `ion` or `ioff`
```bash
cd lib/matplotlib/tests
grep -rni 'ioff()'
```
returns no results. Does this PR warrant adding new tests?

## PR Checklist

- [x] Has Pytest style unit tests 
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
    - See above discussion of documentation
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] NA? Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jklymak commented:
@tacaswell @QuLogic  @efiring   I don't understand `ion/ioff`, but maybe this PR could get some discussion?

QuLogic commented:
Seems fine, but it's still draft...

ianhi commented:
I rebased and started trying to build the docs in order to see how my change affected the docs. I followed the instructions here (https://matplotlib.org/3.1.1/devel/documenting_mpl.html) and even tried them in a fresh conda env:
```
sudo apt install graphviz texlive texlive-xetex
conda create -n mpl-dev python
pip install -e .
pip install -r requirements/doc/doc-requirements.txt
cd docs
make html
```

But I keep running into this mystifying error:
![image](https://user-images.githubusercontent.com/10111092/88758018-75ea9080-d135-11ea-9473-f7bffc621e84.png)

Any advice for fixing that error/discovering what the error is? It makes it much harder to work out the above issues with the docs. Though I can confirm from the [build artifact ](https://42480-1385122-gh.circle-artifacts.com/0/doc/build/html/search.html?q=ioff) that the ioff doc page no longer exists.

ianhi commented:
@QuLogic sorry I didn't see your comment before I posted the above. The code changes work for daily usage (I have started using this by monkeypatching it in and using daily) but the remaining issue is that by making `ioff` and `ion` instances of a class it changes how they show up in the docs and I am not sure if these changes are ok, or if not how to fix them.

ianhi commented:
I just rewrote this context manager in the same manner as the `xkcd` context manager which should return `ion` and `ioff` to being functions which will allow the docs to render correctly (still can't get them to build locally :disappointed: so pushing to see the docs build). I added tests and an api note + new docs and will update the original post's checklist accordingly. If all the tests pass then I think this will be ready for review

ianhi commented:
@jklymak I think this is ready for review. Could you undo the marking as draft (I don't seem to be able to)

New documentation links:
[ioff](https://43065-1385122-gh.circle-artifacts.com/0/doc/build/html/api/_as_gen/matplotlib.pyplot.ioff.html?highlight=ioff#matplotlib.pyplot.ioff)
[ion](https://43065-1385122-gh.circle-artifacts.com/0/doc/build/html/api/_as_gen/matplotlib.pyplot.ion.html?highlight=ion#matplotlib.pyplot.ion)

jklymak commented:
... The "Ready For Review" button doesn't work?  Happy to click it for you but just wanted to see what the interface is?

ianhi commented:
> ... The "Ready For Review" button doesn't work? Happy to click it for you but just wanted to see what the interface is?

:facepalm: I was looking at the top right section of the page - totally the wrong place. Sorry about that and thanks for the tip

jklymak commented:
NP, just didn't want to convert things to draft if the author couldn't un-draft it.....

QuLogic commented on doc/api/next_api_changes/behavior/17371-IHI.rst:
```diff
@@ -0,0 +1,7 @@
+ioff and ion can be used as context managers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
+a context with interactive mode off, or on respectively. The old behavior of
```
The respectively clauses are reversed.
```suggestion
``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
a context with interactive mode on or off, respectively. The old behavior of
```

QuLogic commented on lib/matplotlib/pyplot.py:
```diff
@@ -364,9 +364,41 @@ def isinteractive():
     return matplotlib.is_interactive()
 
 
+class _ioff():
```
```suggestion
class _ioff:
```

QuLogic commented on lib/matplotlib/pyplot.py:
```diff
@@ -364,9 +364,41 @@ def isinteractive():
+        del self.wasinteractive
+
+
+class _ion():
```
```suggestion
class _ion:
```

ianhi commented:
> Does this work recursively? 

Yes it does. Would you like me to add that to tests?

It turns out that by using `call` rather than `init` I was breaking the normal usage of `ion` and `ioff` so I switched to using `init` and updated the tests so that they would catch that. One edge case here is: If you save the return value from a call to `ioff` or `ion` and try to use it as a contextmanager again it will no longer work. 
![image](https://user-images.githubusercontent.com/10111092/89465885-da62ac80-d740-11ea-90cd-80b8c5abb6de.png)

This is pretty niche and hopefully no one is saving the result of `plt.ioff` but perhaps I should add a `_used` variable that is set to true on `__exit__` and then raise a warning in `__enter__`? Like so:

```python
class _ioff:
    _used = False

    def __init__(self):
        self.wasinteractive = isinteractive()
        matplotlib.interactive(False)
        uninstall_repl_displayhook()

    def __enter__(self):
        if self._used:
            cbook._warn_external("re-using an ioff contextmanager is",
                                 " unsupported and may have unexpected results")

    def __exit__(self, exc_type, exc_value, traceback):
        if self.wasinteractive:
            matplotlib.interactive(True)
            install_repl_displayhook()
        self._used = True
```

ianhi commented:
The tests all failed due to a primordial implementation of `_used` sneaking into the commit :/. I can fix that either by removing or implementing the code proposed in the above comment. So I'll wait for feedback re the warning

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -375,25 +406,54 @@ def ioff():
+            # ...
+
+        # This figure will be shown immediately
+        fig2 = plt.figure()
```
I would mention the return value in the notes section for the sake of being explicit.

~~~
    The return value is needed to make the function work as a context
    manager. It is not intended to be stored or accessed by the user.
~~~

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -364,9 +364,40 @@ def isinteractive():
 
 
+class _ioff:
+    def __init__(self):
```
Should get some explanatory docstring. Something along the lines of

~~~
    """
    Context manager for `.ioff`.
    
    The state is changed in ``__init__()`` instead of ``__enter__()``. The
    latter is a no-op. This allows using `.ioff` both as a function and
    as a context.
    """
~~~

ianhi commented:
I added more information to docstrings. Thinking more about the previously proposed `_used` approach I realized it can't be used to protect against using the return value as a contextmanager. What would be necessary is to detect if the `__init__` method was called inside of a `with` context, but that doesn't seem possible. So this pathological case will remain possible:
![image](https://user-images.githubusercontent.com/10111092/89477536-84e7c900-d75b-11ea-8dd3-99d027c80c76.png)


An alternative to adding extra documentation and then hoping that users don't do this we could create a separate `plt.ioff_context` function. This would solve the above issue, but it would add cognitive load to the user. I at least find it more intuitive to call `with plt.ioff():` than `with plt.ioff_context():`. If the tradeoff of not having `ion/ioff` have return values is deemed worthwhile then I'm happy to switch to that approach.


As an aside: What is difference between single and double backticks as used in matplotlib docstrings? This [SO](https://stackoverflow.com/questions/22256995/restructuredtext-in-sphinx-and-docstrings-single-vs-double-back-quotes-or-back) question seems to imply that the meaning of single backticks can vary project to project.

QuLogic commented:
This is reStructuredText; single are cross-references, double are code blocks.

timhoffm commented:
For cross-references see also https://matplotlib.org/devdocs/devel/documenting_mpl.html#referring-to-other-code.

Let me think a little about returning a context manager vs. a separate function.

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -364,9 +364,55 @@ def isinteractive():
     return matplotlib.is_interactive()
 
 
+class _ioff:
```
```suggestion
class _IoffContext:
```

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -375,25 +421,60 @@ def ioff():
+        # This figure will be shown immediately
+        fig2 = plt.figure()
+
+    The return value of `plt.ioff` is necessary to make the function work
+    as a context manager. It is not intended to be stored or accessed
+    by the user.
```
```suggestion
    To enable usage as a context manager, this function returns a
    ``_IoffContext`` object. The return value is not intended to be stored
    or accessed by the user.
```

same for `ion()`.

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -375,25 +421,60 @@ def ioff():
     See Also
     --------
-    ioff : disable interactive mode
+    ion : enable interactive mode
```
This change seems to be a copy-paste error.

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -364,9 +364,55 @@ def isinteractive():
+            install_repl_displayhook()
+
+
+class _ion:
```
```suggestion
class _IoffContext:
```

ianhi commented:
> IMHO it is better to have a function with a (potentially unused) return value context object than creating additional functions ioff/ion_context().

@timhoffm my initial solution avoided the issue of the return value by making `ion/ioff` instances of a class that implemented `__call__`, `__enter__`, and `__exit__`. I've actually been using that myself in the meantime ([ref/code example](https://github.com/ianhi/mpl-interactions/blob/2210e1f67950b16f3c1cd1fb2e5724213c1f71d7/mpl_interactions/utils.py#L13-L28)) but for the life of my I couldn't figure out how to make it work with the documentation as this made `ion/ioff` no longer functions. My initial post on this PR has details on this. But I think that the solution we've ended up at should be just fine :). It is the same approach taken by the `xkcd` context manager and by `rc_context`. 


I've updated the documentation with your helpful clarifications, and I made the examples in `ion` and `ioff` have the same structure.

QuLogic commented on doc/api/next_api_changes/behavior/17371-IHI.rst:
```diff
@@ -0,0 +1,7 @@
+ioff and ion can be used as context managers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
```
This should be cross-referencable now that you've made the change to fix the docs?
```suggestion
`plt.ion()` and `plt.ioff()` may now be used as context managers to create
```

QuLogic commented on doc/api/next_api_changes/behavior/17371-IHI.rst:
```diff
@@ -0,0 +1,7 @@
+
+``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
+a context with interactive mode on, or off respectively. The old behavior of
+calling these functions was maintained. To use the new functionality
```
```suggestion
calling these functions is maintained. To use the new functionality
```

ianhi commented:
I added variations on nested contexts all of which work. I did think of one more pathological case. What is the expected behavior for:
```python
plt.ioff()
with plt.ioff():
    plt.ion()
print(mpl.is_interactive())
```

currently the context manager will leave you in interactive mode at the end

![image](https://user-images.githubusercontent.com/10111092/89857492-3aef5080-db6a-11ea-80b8-30c16e1acbdf.png)


It should be easy to change it so that interactive mode is off in the end, but I'm 50/50 on what's best.

dopplershift commented:
I think the context manager should change state on entry into the block, and restore the previous state on exit. Therefore IMO at the end of your code interactive mode should be off.
